xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
+  // which adds data dependencies between gemms. Instead we write all
+  // intermediate results into a larger buffer and perform a one-shot reduction.
+  // The high-level transformation is:
+  // shape(x,y) cp   shape(x,y) dot0
+  //          \         /
+  //           \       /
+  //     shape(x,y) add0    shape(x,y) dot1
+  //             \                /
+  //              \              /
+  //             shape(x,y) add1
+  //                     |
+  //        shape(x,y) loop output
+  //
+  // transformed into:
+  // shape(x,y) cp         shape(x,y) dot0    shape(x,y) dot1
+  //    \                     /                   /
+  //     \                   /                   /
+  //     shape(n,x,y) concatenate on first axis, n is the number of partitions
+  //                        |
+  //             shape(n,x,y) loop output
+  //                        |
+  //             shape(x,y) reduction on first axis
+  //
+  // The final reduction is pulled outside of the loop to overlap with other
+  // collectives.
+  if (partial_accumulations.size() > 0 &&
+      while_body->name().find(
+          WindowedEinsumHandler::kWindowedEinsumAgLoopName) !=
+          std::string::npos) {
+    std::vector<HloInstruction*> partials_to_concat;
+
+    // We reshape it to a N+1 dimensioned tensor with left-most dim being 1.
+    Shape shape = partial_accumulations[0]->shape();
+    shape = ShapeUtil::PrependMajorDimension(1, shape);
+
+    for (auto& inst : partial_accumulations) {
+      CHECK(inst->user_count() == 1);
+      HloInstruction* reshaped_partial = while_body->AddInstruction(
+          HloInstruction::CreateReshape(shape, inst));
+      partials_to_concat.push_back(reshaped_partial);
+    }
+    Shape concat_shape = partial_accumulations[0]->shape();
+    concat_shape = ShapeUtil::PrependMajorDimension(
+        partial_accumulations.size(), concat_shape);
+
+    HloInstruction* concat = while_body->AddInstruction(
+        HloInstruction::CreateConcatenate(concat_shape, partials_to_concat, 0));
+
+    HloComputation* comp = loop->parent();
+    HloInstruction* windowed_lhs = loop->mutable_operand(0)->mutable_operand(0);
+    // Add a broadcasted zero of the same type as windowed_lhs. This holds all
+    // the partial accumulations and will be fed to a global reduction after
+    // this windowed einsum loop. We move the reduction outside of the loop so
+    // it can be fused or overlap with other instructions in the main
+    // computation.
+    Literal zero_literal =
+        LiteralUtil::Zero(windowed_lhs->shape().element_type());
+    HloInstruction* zero = comp->AddInstruction(
+        HloInstruction::CreateConstant(std::move(zero_literal)));
+    Shape zero_bcast_shape = ShapeUtil::ChangeElementType(
+        concat_shape, windowed_lhs->shape().element_type());
+    HloInstruction* zero_bcast = MakeBroadcastHlo(zero, {}, zero_bcast_shape);
+    loop->mutable_operand(0)->AppendOperand(zero_bcast);
```

Comment:
Check there is only one user of `loop->mutable_operand(0)` and bail if there isn't, as is done [here](https://github.com/openxla/xla/blob/c04ecd9240b648c3a2d9323fc99b8392ccb611ca/xla/service/gpu/transforms/windowed_einsum_handler.cc#L68).

Reply:
makes sense, i will do the check at the beginning of the new function.